testhelpers/output.go

```diff
@@ -0,0 +1,125 @@
+package testhelpers
+
+import (
+	"io"
+	"io/ioutil"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// CmdOutput collects the output from a CLI command issued to a process.
+type CmdOutput struct {
+	// Args is the command and argument sequence executed.
+	Args []string
+
+	outbytes []byte
+	errbytes []byte
+
+	// The status or exit code of a successfully-executed command. This value is only meaningful
+	// after the command completes, and if error is nil.
+	status int
+	// Any error encountered in arranging execution of the command (e.g. I/O errors reading from
+	// streams). An error here indicates that the status and output buffers may not be meaningful.
+	// This is distinguished from an "application" error indicated in the status code and output
+	// streams.
+	error error
+
+	tb testing.TB
+}
+
+// ReadOutput reads the `stdout` and `stderr` streams completely and returns a new Output object.
+func ReadOutput(tb testing.TB, args []string, stdout io.ReadCloser, stderr io.ReadCloser) *CmdOutput {
+	stdoutBytes, err := ioutil.ReadAll(stdout)
+	require.NoError(tb, err)
+
+	stderrBytes, err := ioutil.ReadAll(stderr)
+	require.NoError(tb, err)
```

Comment:
I don't think we will run into this issue, but depending on the size of the buffers used we may deadlock around a write to stderr (logs) as we are producing / reading from stdout.

I don't think we need to fix this right now but wanted to point it out so it's in the back of a few peoples heads just in case we encounter this.

Reply:
Thanks. That seems like a difficult-to-diagnose issue if it were to happen, so I've addressed it now by reading the streams in parallel (potentially consuming lots of memory).